README.md
# edisyn-beatstep

Edisyn service provider to support Arturia Beatstep

FIXME: description

## Installation

Download from https://github.com/danielappelt/edisyn-beatstep/releases.

## Usage

Java JRE version 9 or newer is needed to run the application.

To run the JAR

    java -jar edisyn-beatstep-<version>-standalone.jar

To run from source:

1. Install leiningen from https://leiningen.org
2. Clone this repository
3. `lein do clean, with-profile compile compile, run`

To create a JAR from source:

1. Follow 1. and 2. from above
3. `lein uberjar`

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

## License

Copyright © 2019 Daniel Appelt
Licensed under the Apache License version 2.0

This work is based on synthesizer patch editor [Edisyn](https://github.com/eclab/edisyn) which is included as a git subtree in [src/edisyn](src/edisyn). Edisyn is
Copyright 2017 by Sean Luke
Licensed under the Apache License version 2.0


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Added
- Option to select global MIDI channel
- Categories for encoders and pads

### Changed
- Allow to organize elements of a type in VBoxes
- Use GridLayout to arrange UI elements
- Align Chooser with upstream proposal
- Use Chooser for continuous controllers
- Use Chooser for MIDI channel

### Fixed
- Fix support for sysex params sharing the same key

## [v0.1.1] - 2019-04-16
### Changed
- Exclude unnecessary edisyn artifacts from JARs
- Cleanup code
- Start using this change log

## v0.1.0 - 2019-04-15
### Fixed
- uberjar task now creates a working JAR including the editor

[Unreleased]: https://github.com/danielappelt/edisyn-beatstep/compare/v0.1.1...HEAD
[v0.1.1]: https://github.com/danielappelt/edisyn-beatstep/compare/v0.1.0...v0.1.1


project.clj
(defproject edisyn-beatstep "0.1.1"
  :description "Edisyn service provider to support Arturia Beatstep"
  :url "https://github.com/danielappelt/edisyn-beatstep/"
  :license {:name "Apache License, Version 2.0"
            :url "http://www.apache.org/licenses/LICENSE-2.0"}
  :dependencies [[org.clojure/clojure "1.10.0"]
                 [uk.co.xfactory-librarians/coremidi4j "1.1"]]
  :plugins [[lein-javac-resources "0.1.1"]]
  :hooks [leiningen.javac-resources]
  :source-paths ["src/clojure"]
  :java-source-paths ["src/edisyn"]
  :resource-paths ["resources"]
  :target-path "target/%s"
  :compile-path "%s/classes"
  :aot [edisyn.synth.arturiabeatstep.ArturiaBeatstep]
  :main edisyn.Edisyn
  :profiles {:compile {:resource-paths ^:replace ["compile-resources"]
                       :target-path "target/default"}
             :repl {:resource-paths ^:replace ["compile-resources"]
                    :main nil}
             :uberjar {:aot :all
                       :omit-source true
                       ;; Do not provide compile-time resources in order to
                       ;; avoid a circular conflict.
                       :resource-paths ^:replace []
                       :filespecs [{:type :paths :paths ["resources"]}]
                       :jar-exclusions [#"^(install|jar|libraries|docs)"]
                       :jar-inclusions [#"^docs/manual/Edisyn.pdf"]}})


src/edisyn/edisyn/gui/Chooser.java
/***
    Copyright 2017 by Sean Luke
    Licensed under the Apache License version 2.0
*/

package edisyn.gui;

import edisyn.*;
import java.awt.*;
import java.awt.geom.*;
import javax.swing.border.*;
import javax.swing.*;
import java.awt.event.*;


/**
   A wrapper for JComboBox which edits and responds to changes to a numerical value
   in the model. The numerical value is assumed to of a min/max range 0...n-1,
   which corresponds to the n elements displayed in the JComboBox.  However you can
   change this and in fact map each element to its own special integer.
        
   For the Mac, the JComboBox is made small (JComponent.sizeVariant = small), but this
   probably won't do anything in Linux or Windows.
        
   @author Sean Luke
*/

public class Chooser extends NumericalComponent
    {
    JComboBox combo;
    int addToWidth = 0;

    // The integers corresponding to each element in the JComboBox.
    int[] vals;
    String[] labels;
    ImageIcon[] icons;

    JLabel label = new JLabel("888", SwingConstants.LEFT)
        {
        public Insets getInsets() { return new Insets(0, 0, 0, 0); }
        };

    boolean callActionListener = true;
    
    public String map(int val)
        {
        return "" + combo.getItemAt(val);
        }
    
    public void setCallActionListener(boolean val)
        {
        callActionListener = val;
        }
        
    public void updateBorder()
        {
        super.updateBorder();
        if (combo != null && 
            combo.isEnabled() == synth.isShowingMutation())  // this part prevents us from repeatedly calling setEnabled()... which creates a repaint loop
            {
            if (synth.isShowingMutation())
                combo.setEnabled(false);
            else
                combo.setEnabled(true);
            }
        }

    public void update(String key, Model model) 
        { 
        if (combo == null) return;  // we're not ready yet
                
        int state = getState();
                
        // it's possible that we're sharing a parameter
        // (see for example Blofeld Parameter 9), so here
        // we need to make sure we're within bounds
        int distance = Integer.MAX_VALUE;
        int index = 0;

        // look for the next best value..
        for(int i = 0; i < vals.length; i++)
            if(Math.abs(vals[i] - state) < distance)
                {
                distance = Math.abs(vals[i] - state);
                index = i;
                }

        // This is due to a Java bug.
        // Unlike other widgets (like JCheckBox), JComboBox calls
        // the actionlistener even when you programmatically change
        // its value.  OOPS.
        setCallActionListener(false);
        combo.setSelectedIndex(index);
        setCallActionListener(true);
        }

    public Insets getInsets() 
        { 
        if (Style.CHOOSER_INSETS() == null)
            return super.getInsets();
        else if (Style.isWindows())
            return Style.CHOOSER_WINDOWS_INSETS();
        else
            return Style.CHOOSER_INSETS(); 
        }

    static int[] buildDefaultValues(String[] elements)
        {
        int[] values = new int[elements.length];
        for(int i = 0; i < values.length; i++) values[i] = i;
        return values;
        }

    /** Creates a JComboBox with the given label, modifying the given key in the Style.
        The elements in the box are given by elements, and their corresponding numerical
        values in the model 0...n. */
    public Chooser(String _label, Synth synth, String key, String[] elements)
        {
        this(_label, synth, key, elements, buildDefaultValues(elements));
        }

    /** Creates a JComboBox with the given label, modifying the given key in the Style.
        The elements in the box are given by elements, and their corresponding numerical
        values in the model 0...n. */
    public Chooser(String _label, Synth synth, String key, String[] elements, int[] values)
        {
        this(_label, synth, key, elements, values, null);
        }

    /** Creates a JComboBox with the given label, modifying the given key in the Style.
        The elements in the box are given by elements, with images in icons, and their corresponding numerical
        values in the model 0...n.   Note that OS X won't behave properly with icons larger than about 34 high. */
    public Chooser(String _label, final Synth synth, final String key, String[] elements, ImageIcon[] icons)
        {
        this(_label, synth, key, elements, buildDefaultValues(elements), icons);
        }

    public Chooser(String _label, final Synth synth, final String key, String[] elements, int[] values, ImageIcon[] icons)
        {
        super(synth, key);
                
        label.setFont(Style.SMALL_FONT());
        label.setBackground(Style.BACKGROUND_COLOR()); // TRANSPARENT);
        label.setForeground(Style.TEXT_COLOR());
        //label.setMaximumSize(label.getPreferredSize());

        combo = new JComboBox(elements)
            {
            public Dimension getMinimumSize() 
                {
                return getPreferredSize(); 
                }
            public Dimension getPreferredSize()
                {
                Dimension d = super.getPreferredSize();
                d.width += addToWidth;
                return d;
                }                       

            protected void processMouseEvent(MouseEvent e)
                {
                super.processMouseEvent(e);
                if (e.getID() == MouseEvent.MOUSE_CLICKED)
                    {
                    if (synth.isShowingMutation())
                        {
                        synth.mutationMap.setFree(key, !synth.mutationMap.isFree(key));
                        // wrap the repaint in an invokelater because the dial isn't responding right
                        SwingUtilities.invokeLater(new Runnable() { public void run() { repaint(); } });
                        }
                    }
                }
            };

        combo.putClientProperty("JComponent.sizeVariant", "small");
        combo.setEditable(false);
        combo.setFont(Style.SMALL_FONT());
        combo.setMaximumRowCount(33);           // 33, not 32, to accommodate modulation destinations for Matrix 1000
        
        setElements(_label, elements, values);

        this.icons = icons;
        this.labels = elements;
        if (icons != null)
            {
            combo.setRenderer(new ComboBoxRenderer());
            //if (Style.isMac()) 
            combo.putClientProperty("JComponent.sizeVariant", "regular");
            }

        setState(getState());
                
        setLayout(new BorderLayout());
        add(combo, BorderLayout.CENTER);
        if (isLabelToLeft())
            add(label, BorderLayout.WEST);
        else
            add(label, BorderLayout.NORTH);
        
        
        /// Apparent OS X Java bug: sometimes after you programmatically change
        /// the value of a JComboBox, it no longer sends ActionListener events.  :-(   
        combo.addItemListener(new ItemListener()
            {
            public void itemStateChanged(ItemEvent e)
                {
                // This is due to a Java bug.
                // Unlike other widgets (like JCheckBox), JComboBox calls
                // the actionlistener even when you programmatically change
                // its value.  OOPS.
                if (callActionListener)
                    {
                    setState(vals[combo.getSelectedIndex()]);
                    }
                }
            });
        }
    
    public boolean isLabelToLeft() { return false; }
    
    public void addToWidth(int val)
        {
        addToWidth = val;
        }
              
    public JComboBox getCombo()
        {
        return combo;
        }
        
    public String getElement(int position)
        {
        return (String)(combo.getItemAt(position));
        }
        
    public int getNumElements()
        {
        return combo.getItemCount();
        }
        
    public int getIndex()
        {
        return combo.getSelectedIndex();
        }
        
    public void setIndex(int index)
        {
        setCallActionListener(false);
        combo.setSelectedIndex(index);
        setCallActionListener(true);
        }
        
    public void setLabel(String _label)
        {
        label.setText("  " + _label);
        }
        
    public void setElements(String _label, String[] elements)
        {
        setElements(_label, elements, buildDefaultValues(elements));
        }

    public void setElements(String _label, String[] elements, int[] values)
        {
        setCallActionListener(false);
        label.setText("  " + _label);
        combo.removeAllItems();
        
        for(int i = 0; i < elements.length; i++)
            combo.addItem(elements[i]);

        vals = (int[])values.clone();

        int _min = Integer.MAX_VALUE;
        int _max = Integer.MIN_VALUE;
        for(int i = 0; i < values.length; i++)
            {
            if (_min > values[i]) _min = values[i];
            if (_max < values[i]) _max = values[i];
            }

        setMin(_min);
        setMax(_max);
        setCallActionListener(true);

        combo.setSelectedIndex(0);
        setState(vals[0]);
        revalidate();
        repaint();
        }

    class ComboBoxRenderer extends JLabel implements ListCellRenderer 
        {
        public ComboBoxRenderer() 
            {
            setOpaque(true);
            setHorizontalAlignment(CENTER);
            setVerticalAlignment(CENTER);
            }
 
        /*
         * This method finds the image and text corresponding
         * to the selected value and returns the label, set up
         * to display the text and image.
         */
        public Component getListCellRendererComponent(JList list, Object value, int index, boolean isSelected, boolean cellHasFocus) 
            {
            // Get the selected index. (The index param isn't always valid, so just use the value.)
            //int selectedIndex = ((Integer)value).intValue();
                        
            if (index == -1) index = combo.getSelectedIndex();
            if (index == -1) return this;
                        
            if (isSelected) 
                {
                setBackground(list.getSelectionBackground());
                setForeground(list.getSelectionForeground());
                } 
            else 
                {
                setBackground(list.getBackground());
                setForeground(list.getForeground());
                }
 
            // Set the icon and text.  If icon was null, say so.
            ImageIcon icon = icons[index];
            String label = labels[index];
            setIcon(icon);
            setText(label);
            setFont(list.getFont());
            return this;
            }
        }


    }


src/clojure/edisyn/synth/arturiabeatstep/ArturiaBeatstep.clj
(ns edisyn.synth.arturiabeatstep.ArturiaBeatstep
  (:import [edisyn.gui Category Chooser HBox LabelledDial VBox Style SynthPanel])
  (:gen-class
   :extends edisyn.Synth
   :post-init setup
   :methods [^:static [getSynthName [] String]
             ^:static [recognize [bytes] boolean]
             ^:static [getNumSysexDumpsPerPatch [] int]]
   :prefix "beatstep-"))

(defn beatstep-getSynthName [] "Arturia Beatstep")
(defn beatstep-getHTMLResourceFileName [this] "ArturiaBeatstep.html")
;; TODO
;; (defn beatstep-getDefaultResourceFileName [this] "ArturiaBeatstep.init")

;; value 41 -> Global MIDI Channel
(def channels (reduce #(conj %1 {:label (str (+ %2 1)) :value %2})
                      [{:label "Global Channel" :value 0x41}] (range 16)))
;; See https://de.wikipedia.org/wiki/Musical_Instrument_Digital_Interface#Controller
(def cc-labels
  (let [msb-lsb ["Bank Select", "Modulation", "Breath Controller", "Undefined"
                 "Foot Controller", "Portamento Time", "Data Byte", "Main Volume"
                 "Balance", "Undefined", "Panorama", "Expression"
                 "Effect Control 1", "Effect Control 2", "Undefined", "Undefined"
                 "General Purpose Controller 1", "General Purpose Controller 2", "General Purpose Controller 3", "General Purpose Controller 4"
                 "Undefined", "Undefined", "Undefined", "Undefined"
                 "Undefined", "Undefined", "Undefined", "Undefined"
                 "Undefined", "Undefined", "Undefined", "Undefined"]]
    (into-array
     (map #(str %1 " - " %2)
          (range 128)
          (concat
           (apply concat (for [t ["MSB" "LSB"]] (map #(str % " " t) msb-lsb)))
           ["Hold 1", "Portamento", "Sostenuto", "Soft Pedal"
            "Legato Footswitch", "Hold 2", "Sound Controller 1 (Sound Variation)", "Sound Controller 2 (Harmonic Content)"
            "Sound Controller 3 (Release Time)", "Sound Controller 4 (Attack Time)", "Sound Controller 5 (Brightness)", "Sound Controller 6"
            "Sound Controller 7", "Sound Controller 8", "Sound Controller 9", "Sound Controller 10"
            "General Purpose Controller 5", "General Purpose Controller 6", "General Purpose Controller 7", "General Purpose Controller 8"
            "Portamento Control", "Undefined", "Undefined", "Undefined"
            "Undefined", "Undefined", "Undefined", "Effect 1 Depth"
            "Effect 2 Depth", "Effect 3 Depth", "Effect 4 Depth", "Effect 5 Depth"
            "Data Increment RPN/NRPN", "Data Decrement RPN/NRPN", "NRPN LSB", "NRPN MSB"
            "RPN LSB", "RPN MSB", "Undefined", "Undefined"
            "Undefined", "Undefined", "Undefined", "Undefined"
            "Undefined", "Undefined", "Undefined", "Undefined"
            "Undefined", "Undefined", "Undefined", "Undefined"
            "Undefined", "Undefined", "Undefined", "Undefined"
            "All Sounds Off", "Controller Reset", "Local Control On/Off"
            "All Notes Off",  "Omni Off", "Omni On", "Mono On / Poly Off"
            "Poly On / Mono Off"])))))

(def encoder-accelerations (into-array ["slow" "medium" "fast"]))
;; "full" only creates values 0 or 127
(def velocity-curves (into-array ["linear" "logarithmic" "exponential" "full"]))

(defn create-global [this color]
  (doto (HBox.)
    (.add (Chooser. "Global MIDI Channel" this (str 0x06 "_" 0x40) (into-array (map #(str (+ % 1)) (range 16)))))
    (.add (Chooser. "CV/Gate Channel" this (str 0x0c "_" 0x50)
                    (into-array (map :label channels))
                    (int-array (map :value channels)) nil))
    (.add (Chooser. "Encoder acceleration" this (str 0x04 "_" 0x41) encoder-accelerations))
    (.add (Chooser. "Pad velocity curve" this (str 0x03 "_" 0x41) velocity-curves))))

;; TODO: Add user scale settings
;; Note on:  0a 3d, 0a 39, 0a 31, 0a 21, 0a 01, 0b 3e, 0b 3c, 0b 38, 0b 30, 0b 20, 0b 00
;; Note off: 0a 03, 0a 07, 0a 0f, 0a 1f, 0a 3f, 0b 01, 0b 03, 0b 07, 0b 0f, 0b 1f, 0b 3f

(def encoder-types [{:label "Off", :value 0x00, :comps []}
                    {:label "CC", :value 0x01, :comps [[:channel :cc] :low :high [:behaviour]]}
                    {:label "RPN/NRPN", :value 0x04, :comps [[:channel :coarse] :lsb :msb [:param-type]]}])
;; TODO: low/high values are only relevant for Absolute mode
(def encoder-behaviour (into-array ["Absolute" "Relative (64)" "Relative (0)" "Relative (16)"]))
;; TODO: coarseness determines whether MSB or LSB of NRPN/RPN is sent?
(def encoder-coarseness {"Coarse" 0x06, "Fine" 0x26})
(def encoder-param-type (into-array ["NRPN" "RPN"]))

(defn create-encoder-comps [this index color]
  (let [comps {:channel (Chooser. "MIDI Channel" this (str index "_" 0x02)
                                  (into-array (map :label channels))
                                  (int-array (map :value channels)) nil)
               ;; Please note that, in general, the most restricted component per key will define the key's value range!
               :coarse (Chooser. "Coarse / Fine" this (str index "_" 0x03)
                                 (into-array (keys encoder-coarseness))
                                 (int-array (vals encoder-coarseness)) nil)
               :cc (Chooser. "CC" this (str index "_" 0x03) cc-labels)
               :low (LabelledDial. "Low Value" this (str index "_" 0x04) color 0 127)
               :lsb (LabelledDial. "LSB" this (str index "_" 0x04) color 0 127)
               :high (LabelledDial. "High Value" this (str index "_" 0x05) color 0 127)
               :msb (LabelledDial. "MSB" this (str index "_" 0x05) color 0 127)
               :param-type (Chooser. "Type" this (str index "_" 0x06) encoder-param-type)
               :behaviour (Chooser. "Behaviour" this (str index "_" 0x06) encoder-behaviour)}]
    ;; Update model value restrictions AFTER ui creation
    (.setMin (.getModel this) (str index "_" 0x03) 0)
    (.setMax (.getModel this) (str index "_" 0x03) 127)
    comps))

(def pad-types [{:label "Off", :value 0x00, :comps []}
                {:label "MMC", :value 0x07, :comps [[:command]]}
                {:label "CC", :value 0x08, :comps [[:channel :cc] :on :off [:behaviour]]}
                {:label "Silent CC", :value 0x01, :comps [[:channel :cc] :on :off [:behaviour]]}
                {:label "Note", :value 0x09, :comps [[:channel] :note [:behaviour]]}
                {:label "Program Change", :value 0x0B, :comps [[:channel] :prgchange :lsb :msb]}])
;; We added None/0 in order to have real values start from 1
(def mmc-commands (into-array ["None" "Stop" "Play" "Deferred Play" "FastForward"
                               "Rewind" "Record Strobe" "Record Exit" "Record Ready"
                               "Pause" "Eject" "Chase" "InList Reset"]))

(def pad-behaviour (into-array ["Toggle" "Gate"]))

;; Pad definition. Create pad parameter UIs per pad index
(defn create-pad-comps [this index color]
  {:channel (Chooser. "MIDI Channel" this (str index "_" 0x02)
                      (into-array (map :label channels))
                      (int-array (map :value channels)) nil)
   :command (Chooser. "MMC Command" this (str index "_" 0x03) mmc-commands)
   :cc (Chooser. "CC" this (str index "_" 0x03) cc-labels)
   :note (LabelledDial. "Note" this (str index "_"  0x03) color 0 127)
   :prgchange (LabelledDial. "Program Change" this (str index "_" 0x03) color 0 127)
   :off (LabelledDial. "Off Value" this (str index "_" 0x04) color 0 127)
   :lsb (LabelledDial. "Bank LSB" this (str index "_" 0x04) color 0 127)
   :on (LabelledDial. "On Value" this (str index "_" 0x05) color 0 127)
   :msb (LabelledDial. "Bank MSB" this (str index "_" 0x05) color 0 127)
   :behaviour (Chooser. "Behaviour" this (str index "_" 0x06) pad-behaviour)})

(defn- get-comp [comps entry]
  (if (vector? entry)
    ;; Create a vbox containing entry's contents
    (let [vbox (VBox.)]
      (dorun (for [c entry] (.add vbox (get-comp comps c))))
      vbox)
    ;; If entry is a Chooser, apply it's current selection as model state. This allows us to
    ;; handle separate sysex parameters sharing the same key using separate Chooser instances.
    (let [c (comps entry)]
      (if (= (class c) edisyn.gui.Chooser)
        (.setState c (.getValue c)))
      c)))

;; Display UI depending on chosen type
(defn create-type-ui [this label key types comps]
  (let [comps-box (HBox.)
        chooser (proxy [Chooser] ["Type" this key (into-array (map :label types)) (int-array (map :value types)) nil]
                  (update [key model]
                    ;; Each method fn takes an additional implicit first arg, which is bound to this.
                    (proxy-super update key model)
                    (.removeAll comps-box)
                    (dorun (for [c (:comps (types (.getIndex this)))] (.add comps-box (get-comp comps c))))
                    (.revalidate comps-box)
                    (.repaint this)))]
    (doto (HBox.)
      (.add (doto (VBox.)
              (.add (javax.swing.JLabel. label))
              (.add chooser)))
      (.add comps-box))))

;; Use GridLayout for now: https://docs.oracle.com/javase/tutorial/uiswing/layout/grid.html
(defn create-encoders [this color]
  (let [panel (javax.swing.JPanel. (java.awt.GridLayout. 0 2 5 10))]
    (.setBackground panel (Style/BACKGROUND_COLOR))
    (dorun (for [index (range 16)]
             ;; 0x20-0x2F addresses one of the sixteen encoders
             (.add panel (create-type-ui this (str "Encoder " (+ index 1))
                                         (str (+ 0x20 index) "_" 0x01)
                                         encoder-types
                                         (create-encoder-comps this (+ 0x20 index) color)))))
    panel))

(defn create-pads [this color]
  (let [panel (javax.swing.JPanel. (java.awt.GridLayout. 0 2 5 10))]
    (.setBackground panel (Style/BACKGROUND_COLOR))
    (dorun (for [index (range 16)]
             ;; 0x70-0x7F addresses one of the sixteen pads
             (.add panel (create-type-ui this (str "Pad " (+ index 1))
                                        (str (+ 0x70 index) "_" 0x01)
                                        pad-types
                                        (create-pad-comps this (+ 0x70 index) color)))))
    panel))

(defn create-buttons [this color]
  (let [panel (javax.swing.JPanel. (java.awt.GridLayout. 0 2 5 10))
        labels ["Start" "Stop" "Cntrl/Seq" "Ext. Sync" "Recall" "Store" "Shift" "Chan"]]
    (.setBackground panel (Style/BACKGROUND_COLOR))
    (dorun (for [index (range (count labels))]
             (.add panel (create-type-ui this (labels index)
                                        (str (+ 0x58 index) "_" 0x01)
                                        pad-types
                                        (create-pad-comps this (+ 0x58 index) color)))))
    panel))

(defn beatstep-setup
  "This post init function for the namespace is used to implement the editor's construction."
  [this]
  (let [vbox (VBox.)]
    ;; Not all parameters on global tab seem to be saved in a Beatstep preset. It's more likely
    ;; that Beatstep has three independent patch categories: global, ctrl, seq?!
    (.add vbox (doto (Category. this (beatstep-getSynthName) (Style/COLOR_GLOBAL))
                 (.add (create-global this (Style/COLOR_GLOBAL)))))
    (.add vbox (doto (Category. this "Large encoder" (Style/COLOR_A))
                 (.add (create-type-ui this "Large encoder"
                                       (str 0x30 "_" 0x01)
                                       encoder-types
                                       (create-encoder-comps this 0x30 (Style/COLOR_A))))))
    (.setMin (.getModel this) (str 0x30 "_" 0x03) 0)
    (.setMax (.getModel this) (str 0x30 "_" 0x03) 127)
    (.add vbox (doto (Category. this "Function buttons" (Style/COLOR_A))
                 (.add (create-buttons this (Style/COLOR_A)))))
    (.addTab this "Global" (doto (SynthPanel. this)
                             (.add vbox java.awt.BorderLayout/CENTER))))
  (.addTab this "Encoders" (doto (SynthPanel. this)
                            (.add (doto (VBox.)
                                    (.add (doto (Category. this "Encoders" (Style/COLOR_B))
                                            (.add (create-encoders this (Style/COLOR_B)))) java.awt.BorderLayout/CENTER)))))
  (.addTab this "Pads" (doto (SynthPanel. this)
                         (.add (doto (VBox.)
                                 (.add (doto (Category. this "Pads" (Style/COLOR_C))
                                         (.add (create-pads this (Style/COLOR_C)))) java.awt.BorderLayout/CENTER))))))

;; For the sysex "specifiction" see
;; https://www.untergeek.de/2014/11/taming-arturias-beatstep-sysex-codes-for-programming-via-ipad/
(def sysex-prefix [0xF0 0x00 0x20 0x6B 0x7F 0x42])

;; Get F0 00 20 6B 7F 42 01 00 pp cc F7
(defn sysex-get-param [pp cc]
  ;; (println "get" (format "%02x %02x" pp cc))
  (concat sysex-prefix [0x01 0x00 pp cc 0xF7]))

;; Set F0 00 20 6B 7F 42 02 00 pp cc vv F7 (cc: number of the controller, pp: parameter, vv: value)
(defn sysex-set-param [pp cc vv]
  ;; (println "set" (format "%02x %02x %02x" pp cc vv))
  (concat sysex-prefix [0x02 0x00 pp cc vv 0xF7]))

(defn sysex-recall [number]
  (concat sysex-prefix [0x05 number 0xF7]))

(defn sysex-store [number]
  (concat sysex-prefix [0x06 number 0xF7]))

;; (defn beatstep-getSendsAllParametersInBulk [this] false)
(defn beatstep-getPauseBetweenMIDISends [this] 1)
;; TODO: maybe the following is not needed, or we need to adopt for it elsewhere
;; (defn beatstep-getNumSysexDumpsPerPatch [] (* 16 6))

;; Recognize sysex message bundle making up a complete beatstep patch -> byte[] data
(defn beatstep-recognize [data]
  ;; TODO: Check for BULK data at least using data length. 16 pads * 6 parameters * 12 bytes + x
  ;; (println "recognize:")
  ;; (println (seq data))
  (and (= (count data) (* 16 6 12)) (= (take 6 data) sysex-prefix)))

(defn beatstep-get-param [this key]
  (let [[cc pp] (map #(Integer/valueOf %) (clojure.string/split key #"_"))]
    (sysex-get-param pp cc)))

(defn beatstep-set-param [this key]
  (let [[cc pp] (map #(Integer/valueOf %) (clojure.string/split key #"_"))
        vv (.get (.getModel this) key)]
    (sysex-set-param pp cc vv)))

(defn beatstep-parseParameter [this data]
  ;; Set parameter F0 00 20 6B 7F 42 02 00 pp cc vv F7
  ;; pp parameter, cc pad/encoder, vv value
  ;; (println "parseParameter:")
  ;; (println (seq data))
  (let [[pp cc vv] (drop 8 data)]
    (.set (.getModel this) (str cc "_" pp) (int vv))))

(defn beatstep-parse [this data fromFile]
  (println "parse:")
  (println (seq data))
  ;; TODO dorun?
  (for [d (partition 12 data)] (beatstep-parseParameter this d))
  (.-PARSE_SUCCEEDED this))

(defn beatstep-gatherPatchInfo [this title tempModel writing]
  (let [combo (javax.swing.JComboBox. (into-array (map #(str "Preset " (+ % 1)) (range 15))))
        result (edisyn.Synth/showMultiOption this (into-array '("Preset")) (into-array (vector combo)) 
                                             title "Please select the preset to use.")]
    ;; Please note that internal preset numbers in Beatstep start from 1
    (if result (.set tempModel "number" (+ (.getSelectedIndex combo) 1)))
    result))

(defn beatstep-changePatch [this tempModel]
  ;; Recall patch using param "number" (0..15) from tempModel
  (println (seq (byte-array (sysex-recall (.get tempModel "number")))))
  ;; TODO: maybe evaluate the return value of tryToSendSysex
  (.tryToSendSysex this (byte-array (sysex-recall (.get tempModel "number"))))
  (.simplePause this 10))

;; Equivalent of parseParameter. TODO: we provide the parameter type here so
;; that Clojure can distinguish it from other emit implementations.
(defn beatstep-emit-String [this key]
  (if (not= key "number") (byte-array (beatstep-set-param this key))))

(defn beatstep-emitAll-Model-boolean-boolean [this tempModel toWorkingMemory toFile]
  (object-array
   (map byte-array
        (reduce (fn [res key] (cons (beatstep-set-param this key) res))
                (cons (if tempModel (sysex-store (.get tempModel "number")) '()) '())
                (filter #(not= % "number") (.getKeys (.getModel this)))))))

(defn beatstep-performRequestCurrentDump [this]
  (dorun (for [key (.getKeys (.getModel this)) :while (not= key "number")]
           (.tryToSendSysex this (byte-array (beatstep-get-param this key))))))

(defn beatstep-performRequestDump [this tempModel changePatch]
  ;; We can ignore parameter changePatch here as we always need to change to the
  ;; requested patch first.
  (.performChangePatch this tempModel)
  (beatstep-performRequestCurrentDump this))
